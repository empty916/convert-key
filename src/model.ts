export type KeyMaps = { readonly [k: string]: string };
