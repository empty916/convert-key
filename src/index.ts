
/**
 *
 * @param {*} obj
 */
function isObj(obj: any): obj is Object {
	return typeof obj === 'object' && obj !== null;
}

/**
 * @example
 * import createConvertUtil from 'convert-key';
 * 
 * const keyMaps = {
 * 	a: 'A1',
 * 	b: 'B1',
 * 	f: 'F1'
 * } as const;
 * 
 * const data = {
 * 	a: 1,
 * 	b: {
 * 	    b1: 1,
 * 	    b2: {},
 * 	    b3: '1'
 * 	},
 * 	c: null,
 * 	d: ''
 * }
 * const {convert, revert} = createConvertUtil(keyMaps);
 * const convertData = convert(data);
 * 
 * // convertData is
 * {
 *     A1: number;
 *     B1: {
 *         b1: number;
 *         b2: {};
 *         b3: string;
 *     };
 *     c: null;
 *     d: string;
 * }
 * 
 * // originalData is equal to data
 * const originalData = revert(convertData);
 * @param keyMaps key maps
 * @returns 
 */
const createConvertUtil = <KM extends {readonly [k: string]: string}>(keyMaps: KM) => {
	if (!isObj(keyMaps)) {
		throw new Error('convert-key: keyMaps must be an Object!');
	}
    const convert = <
        DK extends keyof KM,
        D extends Partial<{[k in DK]: any}>
    >(data: D) => {
		if (!isObj(data)) {
			throw new Error('convert-key: data must be an Object!');
		}
        const Res = Object.keys(data).reduce((res, k) => {
            if (keyMaps[k]) {
                // @ts-ignore
                res[keyMaps[k]] = data[k]
            } else {
                // @ts-ignore
                res[k] = data[k]
            }
            return res;
        }, {} as {
            -readonly [PK in keyof KM as Extract<KM[PK], string>]: D[Extract<PK, keyof D>];
        } & {
            [PK in keyof D as Exclude<PK, keyof KM>]: D[PK];
        });
        
        type TR = typeof Res;
        type CK = TR extends Record<infer k, any> ? k : never;

        return Res as {
            [k in CK as TR[k] extends never ? never : k]: TR[k]
        };
    }

    const revertKeyMaps = Object.keys(keyMaps).reduce((res, k) => {
        // @ts-ignore
        res[keyMaps[k]] = k;
        return res;
    }, {} as {
        [k in keyof KM as KM[k]]: k;
    })
    const revert = <
        OK extends keyof KM,
        DK extends KM[OK],
        D extends Partial<{[k in DK]: any}>,
        RKM = {
            [K in OK as KM[K]]: K
        }
    >(data: D) => {
		if (!isObj(data)) {
			throw new Error('convert-key: data must be an Object!');
		}
        const Res = Object.keys(data).reduce((res, k) => {
            // @ts-ignore
            if (revertKeyMaps[k]) {
                // @ts-ignore
                res[revertKeyMaps[k] as any] = data[k]
            } else {
                // @ts-ignore
                res[k] = data[k]
            }
            return res;
        }, {} as {
            [K in DK as Extract<RKM[Extract<K, keyof RKM>], string>]: D[Extract<K, keyof D>];
        } & {
            [PK in keyof D as Exclude<PK, keyof RKM>]: D[PK];
        });

        type TR = typeof Res;
        type CK = TR extends Record<infer k, any> ? k : never;

        return Res as {
            [k in CK as TR[k] extends never ? never : k]: TR[k]
        };
    }
    return {
        convert,
        revert,
    }
};

export default createConvertUtil;