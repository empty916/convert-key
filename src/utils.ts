import { KeyMaps } from "./model";

/**
 *
 * @param {*} obj
 */
export function isObj(obj: any): obj is Object {
  return typeof obj === "object" && obj !== null;
}

export function isObjArray(data: unknown): data is { [s: string]: any }[] {
  if (!Array.isArray(data)) {
    return false;
  }
  if (data.every(isObj)) {
    return true;
  }
  return false;
}

export function getRevertKeyMaps<KM extends KeyMaps>(keyMaps: KM) {
  return Object.keys(keyMaps).reduce(
    (res, k) => {
      // @ts-ignore
      res[keyMaps[k]] = k;
      return res;
    },
    {} as {
      [k in keyof KM as KM[k]]: k extends string ? k : never;
    }
  );
}

/**
 * Copy from d.ts file where is builded from createObjectConvertor.ts
 */
export type ConvertObj<
  KM extends KeyMaps,
  D extends Partial<{ [k in keyof KM]: any }>,
  R = {
    [PK in keyof KM as KM[PK]]: D[Extract<PK, keyof D>];
  } & {
    [PK in keyof D as Exclude<PK, keyof KM>]: D[PK];
  }
> = {
  [k in keyof R as R[k] extends never ? never : k]: R[k];
};


export type ConvertDataDeep<
  KM extends KeyMaps,
  O extends Partial<{ [k in keyof KM]: any }> | Partial<{ [k in keyof KM]: any }>[],
  D = O extends Array<infer I> ? I : O,
  DR = {
    [PK in keyof KM as KM[PK]]: D[Extract<PK, keyof D>];
  } & {
    [PK in keyof D as Exclude<PK, keyof KM>]: D[PK];
  },
  NNR = {
    [k in keyof DR as DR[k] extends never ? never : k]: DR[k];
  }
> = O extends Array<any> ?  ConvertDataDeep<KM, Extract<D, Partial<{ [k in keyof KM]: any }>>>[] : {
  -readonly [K in keyof NNR]: NNR[K] extends Partial<{ [k in keyof KM]: any }> | Partial<{ [k in keyof KM]: any }>[]
    ? ConvertDataDeep<KM, NNR[K]>
    : NNR[K];
};

