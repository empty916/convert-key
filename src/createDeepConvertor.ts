import { KeyMaps } from "./model";
import { ConvertDataDeep, getRevertKeyMaps, isObj, isObjArray } from "./utils";

export const convertDataDeep = <
  KM extends KeyMaps,
  D extends
    | Partial<{ [k in keyof KM]: any }>
    | Array<Partial<{ [k in keyof KM]: any }>>
>(
  keyMaps: KM,
  data: D
): ConvertDataDeep<KM, D> => {
  if (!isObj(data) && !isObjArray(data)) {
    return data;
    // throw new Error("convert-key createDeepConvertor: data must be an Object Or an Object Array!");
  }
  if (Array.isArray(data)) {
    // @ts-ignore
    return data.map((i: D[number]) => convertDataDeep(keyMaps, i));
  }
  return Object.keys(data).reduce((res, k) => {
    let d = data[k];
    if (isObj(d)) {
      d = convertDataDeep(keyMaps, d as D);
    }
    if (keyMaps[k]) {
      // @ts-ignore
      res[keyMaps[k]] = d;
    } else {
      // @ts-ignore
      res[k] = d;
    }
    return res;
  }, {} as ConvertDataDeep<KM, D>);
};

export const createDeepConvertor = <KM extends KeyMaps>(keyMaps: KM) => {
  if (!isObj(keyMaps)) {
    throw new Error("convert-key: keyMaps must be an Object!");
  }
  const convert = <
    D extends
      | Partial<{ [k in keyof KM]: any }>
      | Array<Partial<{ [k in keyof KM]: any }>>
  >(
    data: D
  ) => {
    return convertDataDeep(keyMaps, data);
  };
  const revertKeyMaps = getRevertKeyMaps(keyMaps);
  type RKM = typeof revertKeyMaps;

  const revert = <
    D extends
      | Partial<{ [k in keyof RKM]: any }>
      | Array<Partial<{ [k in keyof RKM]: any }>>
  >(
    data: D
  ) => {
    return convertDataDeep(revertKeyMaps, data);
  };
  return {
    convert,
    revert,
  }
};
