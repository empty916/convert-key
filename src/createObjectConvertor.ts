import { KeyMaps } from "./model";
import { getRevertKeyMaps, isObj } from "./utils";

export const convertObj = <
  KM extends KeyMaps,
  D extends Partial<{ [k in keyof KM]: any }>,
  MK = {
    [K in keyof D]: K extends keyof KM ? KM[K]: K
  },
  OK = {
    [K in keyof MK as Extract<MK[K], string>]: K
  }
>(
  keyMaps: KM,
  data: D
) => {
  if (!isObj(data)) {
    throw new Error("convert-key: data must be an Object!");
  }
  const Res = Object.keys(data).reduce(
    (res, k) => {
      if (keyMaps[k]) {
        // @ts-ignore
        res[keyMaps[k]] = data[k];
      } else {
        // @ts-ignore
        res[k] = data[k];
      }
      return res;
    },
    {} as {
      [PK in keyof KM as KM[PK]]: D[Extract<PK, keyof D>];
    } & {
      [PK in keyof D as Exclude<PK, keyof KM>]: D[PK];
    }
  );

  type TR = typeof Res;
  return Res as unknown as {
    [k in keyof OK]: TR[Extract<k, keyof TR>];
  };
};

/**
 * @example
 * import { createObjectConvertor } from 'convert-key';
 *
 * const keyMaps = {
 * 	a: 'A1',
 * 	b: 'B1',
 * 	f: 'F1'
 * } as const;
 *
 * const data = {
 * 	a: 1,
 * 	b: {
 * 	    b1: 1,
 * 	    b2: {},
 * 	    b3: '1'
 * 	},
 * 	c: null,
 * 	d: ''
 * }
 * const {convert, revert} = createObjectConvertor(keyMaps);
 * const convertData = convert(data);
 *
 * // convertData is
 * {
 *     A1: number;
 *     B1: {
 *         b1: number;
 *         b2: {};
 *         b3: string;
 *     };
 *     c: null;
 *     d: string;
 * }
 *
 * // originalData is equal to data
 * const originalData = revert(convertData);
 * @param keyMaps key maps
 * @returns
 */
export const createObjectConvertor = <KM extends KeyMaps>(keyMaps: KM) => {
  if (!isObj(keyMaps)) {
    throw new Error("convert-key: keyMaps must be an Object!");
  }
  const convert = <D extends Partial<{ [k in keyof KM]: any }>>(data: D) => {
    return convertObj(keyMaps, data);
  };

  const revertKeyMaps = getRevertKeyMaps(keyMaps);

  type RKM = typeof revertKeyMaps;

  const revert = <D extends Partial<{ [k in keyof RKM]: any }>>(data: D) => {
    return convertObj(revertKeyMaps, data);
  };
  return {
    convert,
    revert,
  };
};
