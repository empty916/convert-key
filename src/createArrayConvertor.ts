import { convertObj } from "./createObjectConvertor";
import { KeyMaps } from "./model";
import { getRevertKeyMaps, isObj, isObjArray } from "./utils";

export const createArrayConvertor = <KM extends KeyMaps>(keyMaps: KM) => {
    if (!isObj(keyMaps)) {
		throw new Error('convert-key: keyMaps must be an Object!');
	}

    const convert = <
        A extends Array<Partial<{[k in keyof KM]: any}>>
    >(array: A) => {
        if (!isObjArray(array)) {
			throw new Error('convert-key: createArrayConvertor data must be an Object Array!');
		}
        return array.map((i: A[number]) => convertObj(keyMaps, i));
    }

    const revertKeyMaps = getRevertKeyMaps(keyMaps);
    type RKM = typeof revertKeyMaps;

    const revert = <
        A extends Array<Partial<{[k in keyof RKM]: any}>>
    >(array: A) => {
        if (!isObjArray(array)) {
			throw new Error('convert-key: createArrayConvertor data must be an Object Array!');
		}
        return array.map((i: A[number]) => convertObj(revertKeyMaps, i));
    }

    return {
        convert,
        revert,
    }
}

