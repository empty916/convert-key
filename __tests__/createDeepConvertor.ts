import { createDeepConvertor } from "../src/createDeepConvertor";

const keyMaps = {
  a: "A1",
  b: "B1",
  f: "F1",
} as const; // keyMaps Object must be delared as a const

const data = [
  {
    a: 1,
    b: {
      b: 1,
      b2: {
        b: [
          {
            b: 1,
            b2: {},
            b3: "1",
          },
        ],
      },
      b3: "1",
    },
    c: null,
    d: "",
    // f: {}
  },
];

const data2 = {
  a: 1,
  b: {
    b: 1,
    b2: {
      b: [
        {
          b: 1,
          b2: {},
          b3: "1",
          f: {}
        },
      ],
    },
    b3: "1",
  },
  c: null,
  d: "",
  // f: {}
};



const data3 = {
  a: 1,
  b: {
    b: 1,
    b2: {
      b: [
        {
          b: 1,
          b2: {},
          b3: "1",
          f: {}
        },
      ],
    },
    b3: "1",
    b4: {
      b1: 1,
      b2: {},
      b3: "1",
    },
  },
  c: null,
  d: "",
  // f: {}
};

test("create", () => {
  // @ts-ignore
  expect(() => createDeepConvertor(null)).toThrow();

  const { convert, revert } = createDeepConvertor(keyMaps);

});

test("convert arr", () => {
  const { convert, revert } = createDeepConvertor(keyMaps);

  const cd = convert(data);
  const od = revert(cd);
  expect(cd).not.toEqual(data);
  expect(cd).toMatchSnapshot();
  expect(od).toEqual(data);
});

test("convert obj", () => {
  const { convert, revert } = createDeepConvertor(keyMaps);

  const cd = convert(data2);
  const od = revert(cd);
  expect(cd).not.toEqual(data2);
  expect(cd).toMatchSnapshot();
  expect(od).toEqual(data2);
});

test("convert obj2", () => {
  const { convert, revert } = createDeepConvertor(keyMaps);

  const cd = convert(data3);
  const od = revert(cd);
  expect(cd).not.toEqual(data3);
  expect(cd).toMatchSnapshot();
  expect(od).toEqual(data3);
});

test("convert array of base type", () => {
  const { convert, revert } = createDeepConvertor(keyMaps);

  const data = {
    a: {
      b: 1,
      c: [1,2,3],
      d: ['1', '2', '3'],
      f: [{a: 1}]
    }
  }
  const cd = convert(data);
  const od = revert(cd);
  expect(cd).not.toEqual(data);
  expect(cd).toMatchSnapshot();
  expect(od).toEqual(data);
});
