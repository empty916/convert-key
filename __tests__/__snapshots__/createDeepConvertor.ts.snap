// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert arr 1`] = `
[
  {
    "A1": 1,
    "B1": {
      "B1": 1,
      "b2": {
        "B1": [
          {
            "B1": 1,
            "b2": {},
            "b3": "1",
          },
        ],
      },
      "b3": "1",
    },
    "c": null,
    "d": "",
  },
]
`;

exports[`convert array of base type 1`] = `
{
  "A1": {
    "B1": 1,
    "F1": [
      {
        "A1": 1,
      },
    ],
    "c": [
      1,
      2,
      3,
    ],
    "d": [
      "1",
      "2",
      "3",
    ],
  },
}
`;

exports[`convert obj 1`] = `
{
  "A1": 1,
  "B1": {
    "B1": 1,
    "b2": {
      "B1": [
        {
          "B1": 1,
          "F1": {},
          "b2": {},
          "b3": "1",
        },
      ],
    },
    "b3": "1",
  },
  "c": null,
  "d": "",
}
`;

exports[`convert obj2 1`] = `
{
  "A1": 1,
  "B1": {
    "B1": 1,
    "b2": {
      "B1": [
        {
          "B1": 1,
          "F1": {},
          "b2": {},
          "b3": "1",
        },
      ],
    },
    "b3": "1",
    "b4": {
      "b1": 1,
      "b2": {},
      "b3": "1",
    },
  },
  "c": null,
  "d": "",
}
`;
